File: expo/app/src/flow/use-smile-id-sample-token-binding-rules.ts
import {
  smileIDSampleLiveSession,
  useSmileIDSampleSessionStore,
  type UseSmileIDSampleTokenBindings,
  type UseSmileIDSampleTokenSession,
  type UseSmileIDSampleUserDetailsRequirement,
} from '@smileid/sample-ui';
import type { UseSmileIDValidationException, ValidationState } from '@smileid/usesmileid';

import { useLaunchArgs } from '../use-smile-id-sample-launch';

/// The two refresh scenarios, which keep the fixture tokens.
export const smileIDSampleStartsExpired = (scenario: string): boolean =>
  scenario === 'expiredToken' || scenario === 'badRefresh';

/// The live-session rule in one place.
export const smileIDSampleLiveSessionFor = (
  session: UseSmileIDSampleTokenSession | null,
  scenario: string,
  nowMillis: number,
): UseSmileIDSampleTokenSession | null =>
  smileIDSampleStartsExpired(scenario) ? null : smileIDSampleLiveSession({ live: session }, nowMillis);

/// The bindings a run starting now may read, from the clock rather than the last tick.
export const smileIDSampleLiveBindingsNow = (scenario: string): UseSmileIDSampleTokenBindings | null =>
  smileIDSampleLiveSessionFor(useSmileIDSampleSessionStore.getState().live, scenario, Date.now())?.bindings ??
  null;

/// The bindings a form renders from, re-read on each tick.
export const useSmileIDSampleLiveBindings = (): UseSmileIDSampleTokenBindings | null => {
  const { scenario } = useLaunchArgs();
  const live = useSmileIDSampleSessionStore((state) => state.live);
  const nowMillis = useSmileIDSampleSessionStore((state) => state.nowMillis);
  return smileIDSampleLiveSessionFor(live, scenario, nowMillis)?.bindings ?? null;
};

/// Drops the issues the token already answers; every other rule the SDK applies still stands.
export const smileIDSampleMinusRequirement = (
  state: ValidationState,
  requirement: UseSmileIDSampleUserDetailsRequirement,
): ValidationState => {
  if (state.valid) return state;
  const outstanding = state.issues.filter((issue) => !covers(requirement, issue));
  return outstanding.length === 0 ? { valid: true } : { ...state, issues: outstanding };
};

const covers = (requirement: UseSmileIDSampleUserDetailsRequirement, issue: UseSmileIDValidationException): boolean => {
  const details = issue.errorDetails as { fieldName?: unknown; reason?: unknown } | undefined;
  switch (details?.fieldName) {
    case 'userDetails.givenNames':
      return !requirement.firstName;
    case 'userDetails.lastName':
      return !requirement.lastName;
    // The contact rule is reported against the object; its reason identifies it.
    case 'userDetails':
      return !requirement.contact && String(details.reason).toLowerCase().includes('email');
    default:
      return false;
  }
};

File: flutter/app/lib/src/flow/use_smileid_sample_token_binding_rules.dart
import 'package:sample_ui/sample_ui.dart';
import 'package:usesmileid/usesmileid.dart';

/// The two scenarios that are about refresh, which a scanned token has no journey for.
bool useSmileIDSampleStartsExpired(UseSmileIDSampleScenario scenario) =>
    scenario == UseSmileIDSampleScenario.expiredToken ||
    scenario == UseSmileIDSampleScenario.badRefresh;

/// The live-session rule in one place, so nothing disagrees about whether a token is live.
UseSmileIDSampleTokenSession? useSmileIDSampleLiveSession(
  UseSmileIDSampleTokenSession? session,
  UseSmileIDSampleScenario scenario,
  int nowMillis,
) =>
    session != null &&
        !session.hasExpired(nowMillis) &&
        !useSmileIDSampleStartsExpired(scenario)
    ? session
    : null;

/// Drops the issues the token already answers; every other rule the SDK applies still stands.
ValidationState useSmileIDSampleMinusRequirement(
  ValidationState state,
  UseSmileIDSampleUserDetailsRequirement requirement,
) {
  if (state is! ValidationStateInvalid) {
    return state;
  }
  final List<UseSmileIDValidationException> outstanding =
      <UseSmileIDValidationException>[
        for (final UseSmileIDValidationException issue in state.issues)
          if (!_covers(requirement, issue)) issue,
      ];
  return outstanding.isEmpty
      ? const ValidationStateValid()
      : ValidationStateInvalid(outstanding);
}

bool _covers(
  UseSmileIDSampleUserDetailsRequirement requirement,
  UseSmileIDValidationException issue,
) {
  if (issue is! InvalidFieldValueException) {
    return false;
  }
  return switch (issue.fieldName) {
    'userDetails.givenNames' => !requirement.firstName,
    'userDetails.lastName' => !requirement.lastName,
    // The contact rule is reported against the object, so its reason is what identifies it.
    'userDetails' =>
      !requirement.contact && issue.reason.toLowerCase().contains('email'),
    _ => false,
  };
}
